core/commands/block.go

```diff
@@ -185,3 +188,99 @@ func getBlockForKey(req cmds.Request, skey string) (blocks.Block, error) {
 	log.Debugf("ipfs block: got block with key: %q", b.Key())
 	return b, nil
 }
+
+var blockRmCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Remove IPFS block(s).",
+		ShortDescription: `
+'ipfs block put' is a plumbing command for removing raw ipfs blocks.
+It takes a list of base58 encoded multihashs to remove.
+`,
+	},
+	Arguments: []cmds.Argument{
+		cmds.StringArg("hash", true, true, "Bash58 encoded multihash of block(s) to remove."),
+	},
+	Options: []cmds.Option{
+		cmds.BoolOption("ignore-pins", "Ignore pins.").Default(false),
+	},
+	Run: func(req cmds.Request, res cmds.Response) {
+		ignorePins, _, err := req.Option("ignore-pins").Bool()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+		n, err := req.InvocContext().GetNode()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+		hashes := req.Arguments()
+		keys := make([]key.Key, 0, len(hashes))
+		for _, hash := range hashes {
+			k := key.B58KeyDecode(hash)
+			keys = append(keys, k)
+		}
+		rdr, wtr := io.Pipe()
+		go func() {
+			pinning := n.Pinning
+			if ignorePins {
+				pinning = nil
+			}
+			err := rmBlocks(n.Blockstore, pinning, wtr, keys)
+			if err != nil {
+				wtr.CloseWithError(fmt.Errorf("Some blocks not deleted: %s", err))
+			} else {
+				wtr.Close()
+			}
+		}()
+		res.SetOutput(rdr)
+		return
+	},
+	Marshalers: cmds.MarshalerMap{
+		cmds.Text: func(res cmds.Response) (io.Reader, error) {
+			return res.(io.Reader), nil
+		},
+	},
+}
+
+type noopLocker struct{}
+
+func (noopLocker) Unlock() {}
+
+// pins may be nil
+func rmBlocks(blocks bs.GCBlockstore, pins pin.Pinner, out io.Writer, keys []key.Key) error {
+	var unlocker bs.Unlocker = noopLocker{}
+	defer unlocker.Unlock()
```

Comment:
this reference will be bound such that it will always be called on the noop lock, even when the value in the unlocker variable is changed. If you want to do things this way, you will need to capture the variable in a closure:
```
defer func() {
    unlocker.Unlock()
}()
```